File: arrays/01_flood_fill/flood_fill_test.cpp
#include "flood_fill.cpp"
#include <gtest/gtest.h>
#include <vector>
 
using namespace std;


class FloodFillTest : public ::testing::Test {
protected:
    Solution* solution; // Each test will have its own pointer to Solution instance

    void SetUp() override {
        // Code here will be called immediately after the constructor 
        // (right before each test).
        solution = new Solution();
    }

    void TearDown() override {
        // Code here will be called immediately after each test 
        // (right before the destructor).
        delete solution;
    }
};


TEST_F(FloodFillTest, SimpleCase) { 
    vector<vector<int>> image = {{1, 1, 1}, {1, 1, 0}, {1, 0, 1}};
    int sr = 1, sc = 1, newColor = 2;
    vector<vector<int>> expected = {{2, 2, 2}, {2, 2, 0}, {2, 0, 1}};

    ASSERT_EQ(expected, solution->floodFill(image, sr, sc, newColor));
}

TEST_F(FloodFillTest, SameStartColor) { 
    vector<vector<int>> image = {{0, 1, 2}, {3, 4, 5}};
    int sr = 0, sc = 0, newColor = 0;
    vector<vector<int>> expected = {{0, 1, 2}, {3, 4, 5}};

    ASSERT_EQ(expected, solution->floodFill(image, sr, sc, newColor));
}

TEST_F(FloodFillTest, ComplexCase) { 
    vector<vector<int>> image = {{6, 5, 1}, {2, 1, 1}, {1, 1, 3}, {1, 0, 1}};
    int sr = 2, sc = 1, newColor = 7;
    vector<vector<int>> expected = {{6, 5, 7}, {2, 7, 7}, {7, 7, 3}, {7, 0, 1}};

    ASSERT_EQ(expected, solution->floodFill(image, sr, sc, newColor));
}


int main(int argc, char **argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: arrays/01_flood_fill/flood_fill_test.py
from unittest import TestCase, main
from flood_fill import Solution
import numpy as np


class TestFloodFill(TestCase):
    def setUp(self):
        # This method is called before each test
        self.solution = Solution()

    def test_flood_fill_simple_case(self):
        image = np.array([[1, 1, 1], [1, 1, 0], [1, 0, 1]])
        sr, sc, new_color = 1, 1, 2
        expected = np.array([[2, 2, 2], [2, 2, 0], [2, 0, 1]])
        result = self.solution.flood_fill(image, sr, sc, new_color)
        np.testing.assert_array_equal(result, expected)

    def test_flood_fill_same_start_color(self):
        image = np.array([[0, 1, 2], [3, 4, 5]])
        sr, sc, new_color = 0, 0, 0
        expected = np.array([[0, 1, 2], [3, 4, 5]])
        result = self.solution.flood_fill(image, sr, sc, new_color)
        np.testing.assert_array_equal(result, expected)

    def test_flood_fill_complex_case(self):
        image = np.array([[6, 5, 1], [2, 1, 1], [1, 1, 3], [1, 0, 1]])
        sr, sc, new_color = 2, 1, 7
        expected = np.array([[6, 5, 7], [2, 7, 7], [7, 7, 3], [7, 0, 1]])
        result = self.solution.flood_fill(image, sr, sc, new_color)
        np.testing.assert_array_equal(result, expected)

    def tearDown(self):
        del self.solution  # Ensure the reference is deleted.


if __name__ == '__main__':
    main()
